lib/tasks/publish_provider_check.rake

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+class PublishProviderChecker
+  include ServicePattern
+
+  PUBLISH_PROVIDER_ENDPOINT = "https://api.publish-teacher-training-courses.service.gov.uk/api/public/v1/recruitment_cycles/2024/providers"
+
+  attr_reader :provider_matches, :lead_school_matches, :lead_partner_matches, :missing
+
+  def initialize
+    @provider_matches = []
+    @lead_school_matches = []
+    @lead_partner_matches = []
+    @missing = []
+  end
+
+  def call
+    next_link = PUBLISH_PROVIDER_ENDPOINT
+    while next_link.present?
+      response = HTTParty.get(next_link).parsed_response
+      response["data"].map { |p| p["attributes"] }.each do |provider|
+        if lead_school_matches?(provider)
+          lead_school_matches << provider
+        elsif lead_partner_matches?(provider)
+          lead_partner_matches << provider
+        elsif provider_matches?(provider)
+          provider_matches << provider
+        else
+          missing << provider
+        end
+      end
+      next_link = response["links"]["next"]
+    end
+
+    self
+  end
+
+  def total_count
+    lead_school_matches.count + lead_partner_matches.count + provider_matches.count + missing.count
+  end
+
+private
+
+  def provider_matches?(provider)
+    Provider.find_by(code: provider["code"]).present? || Provider.find_by(ukprn: provider["ukprn"]).present?
+  end
+
+  def lead_school_matches?(provider)
+    School.find_by(urn: provider["urn"]).present?
+  end
+
+  def lead_partner_matches?(provider)
+    LeadPartner.find_by(urn: provider["urn"]).present? || LeadPartner.find_by(ukprn: provider["ukprn"]).present?
+  end
+end
+
+class PublishProviderReporter
```

Comment:
maybe class per file...

This looks like it can perhaps be inline into the rake as most bits are just console printout